R/Quosure.R

```diff
@@ -0,0 +1,307 @@
+#' Reproducible class with environment and code.
+#'
+#' @slot code (`character`) to reproduce the environment
+#' @slot env (`environment`) environment which content was generated by the evaluation
+#' of the `code` slot.
+setClass(
+  "Quosure",
+  representation(code = "character", env = "environment"),
+  prototype(code = character(0), env = new.env(parent = parent.env(.GlobalEnv)))
+)
+
+#' Initialize `Quosure` object
+#'
+#' Initialize `Quosure` object with `code` and `env`. In order to have `Quosure` reproducible
+#' one needs to initialize with `env` which can be reproduced by the `code`. Alternatively, one
+#' can create an empty `Quosure` and evaluate the expressions in this object.
+#' @name new_quosure
+#'
+#' @param code (`character(1)` or `language`) code to evaluate. Also acceptps and stores comments
+#' @param env (`environment`) name of the code element.
+#'
+#' @export
+setGeneric("new_quosure", function(code = character(0), env = parent.env(.GlobalEnv)) {
+  standardGeneric("new_quosure")
+})
+
+#' @rdname new_quosure
+#' @export
+setMethod(
+  "new_quosure", signature(code = "missing", env = "missing"),
+  function(code, env) {
+    methods::new("Quosure")
+  }
+)
+
+#' @rdname new_quosure
+#' @export
+setMethod(
+  "new_quosure", signature(code = "character", env = "environment"),
+  function(code = character(0), env = parent.env(.GlobalEnv)) {
+    checkmate::check_class(env, "environment")
+    checkmate::check_string(code)
+
+    lockEnvironment(env)
+
+    if (is.null(names(code))) {
+      names(code) <- "code"
+    }
+    methods::new("Quosure", code = code, env = env)
+  }
+)
+
+
+#' @rdname new_quosure
+#' @export
+setMethod(
+  "new_quosure", signature(code = "character", env = "list"),
+  function(code, env) {
+    env <- if (checkmate::check_list(env, "reactive")) {
+      lapply(env, function(x) {
+        if (inherits(x, "reactive")) {
+          x()
+        } else {
+          x
+        }
+      })
+    }
+
+    new_quosure(code = code, env = list2env(env))
+  }
+)
+
+
+setGeneric("get_code", function(object, deparse = FALSE) {
+  standardGeneric("get_code")
+})
+
+setGeneric("join", function(object, object2, overwrite = FALSE) {
+  standardGeneric("join")
+})
+
+#' Evaluate the code in the `Quosure` environment
+#'
+#' Given code is evaluated in the `Quosure` environment and appended to the `code` slot. This means
```

Comment:
> and appended to the `code` slot

What if we want to store different code that the one evaluated? I am thinking about obfuscation of passwords etc.

Reply:
I don't have an idea how to enable this. I think in DDL we should rather do:
```r
expr <- rice_connect(login = login_val, password = password_val)
expr_to_eval <- substitute(expr, list(login = input$login, password = input$password))
expr_to_return <- substitute(expr, list(login = Sys.getvar("USER"), password = quote(askpass::askpass()))
new_quosure(eval(expr), code = expr_to_return))
```

Don't you think that expr replacement is little to much to the Quosure?

Reply:
I like your idea and I agree Quosure shouldn't be responsible for expr replacement. It needs to be done on client side (inside the module) or... a separate class that Quosure could use.

```
expr <- rice_connect(...)
expr_obj <- obfuscate_substitute(expr, list(path = input$path, login = input$login, password = input$password), list(login = ..., password = ...)
new_quosure(eval(expr_obj$code), expr_obj$obfuscated_code)
# or alternatively
new_quosure(expr_obj)
```

All in all, quosure API needs to allow passing those two values - either combined into the single object or provided separately.